index.js

```diff
@@ -0,0 +1,153 @@
+
+var util = require('util'),
+    winston = require('winston'),
+    AWS = require('aws-sdk'),
+    dynamodbIntegration = require('./lib/dynamodb-integration'),
+    isEmpty = require('lodash.isempty'),
+    assign = require('lodash.assign'),
+    isError = require('lodash.iserror'),
+    stringify = require('./lib/utils').stringify,
+    debug = require('./lib/utils').debug,
+    defaultFlushTimeoutMs = 10000;
+
+const ALLOWED_REGIONS = ['us-east-1', 'us-west-1', 'us-west-2', 'eu-west-1', 'eu-central-1', 'ap-northeast-1', 'ap-northeast-2', 'ap-southeast-1', 'ap-southeast-2', 'sa-east-1'];
+
+var WinstonDynamoDB = function (options) {
+    winston.Transport.call(this, options);
+    this.level = options.level || 'info';
+    this.name = options.name || 'DynamoDB';
+    this.tableName = options.tableName;
+    this.logStreamName = options.logStreamName;
+    this.options = options;
+
+    var awsAccessKeyId = options.awsAccessKeyId;
+    var awsSecretKey = options.awsSecretKey;
+    var awsRegion = options.awsRegion;
+    var messageFormatter = options.messageFormatter ? options.messageFormatter : function(log) {
+        return [ log.level, log.message ].join(' - ')
+    };
+    this.formatMessage = options.jsonMessage ? stringify : messageFormatter;
+    this.proxyServer = options.proxyServer;
+    this.uploadRate = options.uploadRate || 2000;
+    this.logEvents = [];
+    this.errorHandler = options.errorHandler;
+
+    if (options.dynamoDB) {
+        this.dynamoDB = options.dynamoDB;
+    } else {
+        if (this.proxyServer) {
+            AWS.config.update({
+                httpOptions: {
+                    agent: require('proxy-agent')(this.proxyServer)
+                }
+            });
+        }
+
+        var config = {};
+
+        if (awsAccessKeyId && awsSecretKey && awsRegion) {
+            config = {credentials: {accessKeyId: awsAccessKeyId, secretAccessKey: awsSecretKey}, region: awsRegion};
+        } else if (awsRegion && !awsAccessKeyId && !awsSecretKey) {
+            // Amazon SDK will automatically pull access credentials
+            // from IAM Role when running on EC2 but region still
+            // needs to be configured
+            config = { region: awsRegion };
+        }
+
+        if (options.awsOptions) {
+            config = assign(config, options.awsOptions);
+        }
+
+        this.dynamoDB = new AWS.DynamoDB(config);
+    }
+
+    if (ALLOWED_REGIONS.indexOf(this.dynamoDB.config.region) < 0) {
+        throw new Error("unavailable region given");
+    }
+
+    debug('constructor finished');
+};
+
+util.inherits(WinstonDynamoDB, winston.Transport);
+
+WinstonDynamoDB.prototype.log = function (info, callback) {
+    debug('log (called by winston)', info);
+
+    if (!isEmpty(info.message) || isError(info.message)) {
+        this.add(info);
+    }
+
+    if (!/^uncaughtException: /.test(info.message)) {
+        // do not wait, just return right away
+        return callback(null, true);
+    }
```

Comment:
Since I'm not too sure on this code here (Specifically what the `callback(null, true) means`. It seems that the callback is called immediately with these params on `logger.info`, but is called again with other params when the batch is sent), I'll probably skip this...

Reply:
callbacks usually follow the pattern of `func(error, result)` and in this case `cb(null, true)` means everything worked. This is the 'happy flow' and will happen for EVERY log call except when the message includes `uncaughtException`. 

It's confusing because the 'normal' case is inside the if and short circuits this function.

If the error message does include `uncaughtException` then we clear everything and submit the logs and get ready for a 'crash' exit of the process.